Adds project list descriptions generation

dominoFire opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/119
Regression: No
* Last working version: N/A
* How are we preventing it in future: 

## Fix

Adds a MSBuild task to generate a markdown file with the project descriptions

## Testing/Validation

Tests Added: No  
Reason for not adding tests: Engineering infra.
Validation: Manual run

Commit:
***NO_CI*** WiP for generating doc overview

Commit:
Wiring up MSBuild tasks

Commit:
First working prototype

Commit:
***NO_CI*** Generated docs in artifacts folder

Commit:
***NO_CI*** Generated timestamp

Commit:
***NO_CI*** Removed XmlPeek task

Commit:
Docs generation described in contribution guide

Commit:
Updated projects item group

zkat reviewed (approved):
yay! This is great!

nkolev92 commented on CONTRIBUTING.md:
```diff
@@ -93,6 +93,7 @@ NuGet members may contribute directly to the main remote.
 
 - `$(NuGetClientRoot)\artifacts\VS15` - this folder will contain the Package Manager extension (`NuGet.Tools.vsix`) and NuGet command-line client application (`NuGet.exe`)
 - `$(NuGetClientRoot)\artifacts\nupkgs` - this folder will contain all our projects packages
+- `$(NuGetClientRoot)\artifacts\docs-generated` - Containes generated docs. Created with `msbuild .\build\build.proj /t:GenerateDocs`
```
I think the docs should be generated into the docs folder at the repo root. 

Is there a good reason why we'd an intermediate doc?

dominoFire replied:
Code-generated artifacts normally are not checked-in in a source control repo. I will add the generated artifact.

nkolev92 replied:
I don't get what you mean.
The doc that this tool is supposed to generate can be a part of the PR. 
It's not a "code-generated" artifact comparable to dlls etc.

dominoFire replied:
Moved the project-list.md to docs\ folder

nkolev92 commented on build/build.proj:
```diff
@@ -415,4 +415,28 @@
+  <Target Name="GenerateDocs">
+    <MsBuild
+      Projects="$(RepositoryRootDirectory)\tools-local\doc.tasks\doc.tasks.csproj"
+      Targets="Restore;Build"
```
Don't do this. 
See: https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restoring-and-building-with-one-msbuild-command.

dominoFire replied:
Changed to two MSBuild task invocations

nkolev92 commented on tools-local/doc.tasks/GenerateMarkdownDoc.cs:
```diff
@@ -0,0 +1,93 @@
+            Directory.CreateDirectory(dirname);
+            Log.LogMessage(MessageImportance.Low, $"Created directories: {dirname}");
+
+            Log.LogMessage(MessageImportance.Low, $"Creating documentation: {outputFilePath}");
```
Maybe it's a good idea to write a basically intro to the overview and then have 2 categories of projects in the list. 
Specifically have NuGet.Clients and NuGet.Core.
But that can be done later. 
We can make incremental changes.

dominoFire replied:
Added tests projects to the overview.

dominoFire replied:
Added test projects to the list.

dominoFire replied:
Also, I added a brief description in project-list.md

nkolev92 commented on tools-local/doc.tasks/GenerateMarkdownDoc.cs:
```diff
@@ -0,0 +1,93 @@
+
+                file.WriteLine($"# {Title}\n\n");
+
+                for (int i = 0; i < ProjectFiles.Length; i++)
```
Sort the ProjectFiles by name.

dominoFire replied:
Sorted.

nkolev92 reviewed (commented):
Please move the generated doc to docs/*

It'll be great if you can add a sample of the generated doc in this PR.

dominoFire committed:
Addressed feedback

dominoFire committed:
Added generated docs

dominoFire committed:
Updated docs location

nkolev92 commented:
@dominoFire Please re-request review when ready. 

The status was a bit ambiguous to me based on your last comment :)

dominoFire committed:
Test without namespaces

dominoFire committed:
Integrating new descriptions

- Addressing previous feedback
- Created new project file for docs generation
- Added project count
- Updated generated project list

nkolev92 commented on docs/project-list.md:
```diff
@@ -0,0 +1,103 @@
+
+Projects count: 37
+
+- [`src\NuGet.Clients\NuGet.CommandLine\NuGet.CommandLine.csproj`](../src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj): NuGet Command Line Interface.
```
How about we use the project name as the link name? 
Right now it looks kind of clunky.

dominoFire replied:
Changed to project name.

nkolev92 commented on docs/project-list.md:
```diff
@@ -0,0 +1,103 @@
+# NuGet Project Overview
+
+Below is a list of all source code projects for NuGet libraries and supported NuGet clients
+
```
Lots of unnecessary empty lines here :)

dominoFire replied:
Removed some empty lines. Added double empty lines between sections. Do you want exactly one empty line between sections ?

nkolev92 commented on docs/project-list.md:
```diff
@@ -0,0 +1,103 @@
+---
+
+
+# NuGet Project Overview
```
I'd also call the file "project-overview.md"

dominoFire replied:
Changed to project-overview.md

nkolev92 commented on tools-local/doc.tasks/GenerateMarkdownDoc.cs:
```diff
@@ -0,0 +1,127 @@
+                file.WriteLine("---");
+
+                file.WriteLine($"\n\n# {Title}\n");
+                file.WriteLine("Below is a list of all source code projects for NuGet libraries and supported NuGet clients\n");
```
This feels redundant.

dominoFire replied:
Reprhased summary

heng-liu commented on tools-local/doc.tasks/GenerateMarkdownDoc.cs:
```diff
@@ -0,0 +1,127 @@
+                file.WriteLine($"\n\n## {TestProjectsSubtitle}\n");
+                file.WriteLine("Most production assemblies has an associated test project, whose name ends with `.Test`.\n");
+                file.WriteLine($"Test Projects count: {TestProjects.Length}\n");
+                SortAndWriteDescriptions(TestProjects, file);
```
Shall we have a separate test-overview.md file for test projects?

dominoFire replied:
We can have a separte file. Now, the task supports generating two tasks. Please, review again.

nkolev92 replied:
cc @donnie-msft thoughts on this. 
https://github.com/NuGet/Client.Engineering/issues/239 will be impacted by that decision.

heng-liu replied:
As we discussed offline, we can have a separate test-overview.md file.

donnie-msft replied:
What is the rationale for having separate files?

Commit:
Addresssed feedback

- Added Section description in item group
- Project file as item bullet
- Relative paths

Commit:
Fixed VS hang when loading doc.task.csproj

nkolev92 commented on docs/project-overview.md:
```diff
@@ -0,0 +1,96 @@
+---
```
You are missing the func tests.

dominoFire replied:
😅 Added functional tests section.

nkolev92 commented on docs/project-overview.md:
```diff
@@ -0,0 +1,96 @@
+
+# NuGet Project Overview
+
+Below is a list of projects contained in the NuGet.Client repo, organized by source and test projects
```
Use punctuation at the end.

nkolev92 replied:
> source and test projects

below you call them product projects...i think that's a better name, so maybe use it here if you agree?

dominoFire replied:
Agree. Changed to `...by product and test projects.`.

nkolev92 reviewed (approved):
Looks good, however, you are the missing the functional tests.

Approving now...just make sure you've included all projects.

Commit:
Added functional tests

Commit:
Format
